README.md
django-future-url
=================

Migration tool for old style “url” tags.
“In Django 1.5, the behavior of the url template tag will change”
https://docs.djangoproject.com/en/1.4/ref/templates/builtins/#url

If you use old style url tags in django 1.4 you will see that:
“DeprecationWarning: The syntax for the url template tag is changing.
Load the ``url`` tag from the ``future`` tag library to start using the
new behavior.”

How it works
------------

-  Finds all \*.html, \*.txt files.
-  Replaces all old style “url” tags.
-  Inserts “{% load url from future %}” when it's needed.

Installation
------------

::
    python setup.py install

Usage
-----

::
    cd ~/projects/my_django_14_project/
    django-make-future-url.py

It will update all templates in directory
“~/projects/my\_django\_14\_project/” and subdirectories.


Options
~~~~~~~

*You can use ``django-make-future-url.py --help`` for help.*

Only shows changes to be made without actually modifying files:
``bash django-make-future-url.py --dry-run``

Verbose output: ``bash django-make-future-url.py --verbose``

Example
-------

``cat ./template1``

    {% url path.to.view arg arg2 %}
    {%  url path.to.view arg arg2 %}
    {%url myapp:view-name %}
    {% url path.to.view as the_url%}
    {% url   path.to.view arg arg2   as   the_url %}
    {%url app_views.client client.id%}

``django-make-future-url.py``

``cat ./template1``

    {% load url from future %}
    {% url 'path.to.view' arg arg2 %}
    {%  url 'path.to.view' arg arg2 %}
    {%url 'myapp:view-name' %}
    {% url 'path.to.view' as the_url%}
    {% url   'path.to.view' arg arg2   as   the_url %}

CHANGELOG.rst
Changelog
=========

0.2.2 (2013-01-28)
^^^^^^^^^^^^^^^^^^
* Clarification for README


0.2.1 (2013-01-28)
^^^^^^^^^^^^^^^^^^
* Fixed setup.py to properly define runner
* Added even more tests
* Improved output
* Made ``--verbose`` a default behavior


0.2 (2013-01-27)
^^^^^^^^^^^^^^^^
* Renamed the tool from ``django-make-future-url.py`` to  ``future_url``
* Removed ``--dry-run`` option, to make actual changes use ``--write``
* Added python 3 support (Roman Gladkov)
* Internal refactoring
* Added tests


0.1 (2012-04-24)
^^^^^^^^^^^^^^^^
* Initial release



tests/test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''Unit tests for module.'''

import unittest


def main():
    unittest.main()

if __name__ == '__main__':
    main()

django_future_url/test/test_core.py
import codecs
import os
import subprocess
import unittest
from django_future_url.core import parse_file


class ModernizeUrlTagTestCase(unittest.TestCase):

    def assertTemplateFixed(self, template_name):
        base = os.path.abspath(os.path.dirname(__file__))
        input_filepath = os.path.join(base, 'examples', template_name + '.html')
        expected_filepath = os.path.join(base, 'expected', template_name + '.fixed.html')

        output = parse_file(codecs.open(input_filepath, 'r', 'utf-8').read())
        expected_output = codecs.open(expected_filepath, 'r', 'utf-8').read()
        self.assertEqual(output, expected_output)


class SimpleTest(ModernizeUrlTagTestCase):

    def test_deprecated_urls(self):
        self.assertTemplateFixed('deprecated_urls')

    def test_deprecated_urls_extends(self):
        self.assertTemplateFixed('deprecated_urls_extends')

    def test_deprecated_urls_extends2(self):
        self.assertTemplateFixed('deprecated_urls_extends2')

    def test_deprecated_urls_false_load(self):
        self.assertTemplateFixed('deprecated_urls_false_load')

    def test_should_not_be_upgraded(self):
        self.assertTemplateFixed('should_not_be_upgraded')

django_future_url/test/test_django_make_future_url.py
import codecs
import os
import unittest
from django_future_url.core import parse_file


class ModernizeUrlTagTestCase(unittest.TestCase):

    def assertTemplateFixed(self, template_name):
        base = os.path.abspath(os.path.dirname(__file__))
        input_filepath = os.path.join(base, 'examples', template_name + '.html')
        expected_filepath = os.path.join(base, 'examples', template_name + '.fixed.html')

        output = parse_file(codecs.open(input_filepath, 'r', 'utf-8').read())
        expected_output = codecs.open(expected_filepath, 'r', 'utf-8').read()
        self.assertEqual(output, expected_output)


class SimpleTest(ModernizeUrlTagTestCase):

    def test_deprecated_urls(self):
        self.assertTemplateFixed('deprecated_urls')

    def test_deprecated_urls_extends(self):
        self.assertTemplateFixed('deprecated_urls_extends')

    def test_deprecated_urls_extends2(self):
        self.assertTemplateFixed('deprecated_urls_extends2')

    def test_deprecated_urls_false_load(self):
        self.assertTemplateFixed('deprecated_urls_false_load')

    def test_should_not_be_upgraded(self):
        self.assertTemplateFixed('should_not_be_upgraded')

django_future_url/core.py
# coding: utf-8

""" Searches for templates in current directory recursively and
modernizes them to use new {% url %} syntax.
Adds {% load url from future %} on top of your template if necessary (Django 1.3-1.4).

From the Django docs
https://docs.djangoproject.com/en/1.4/ref/templates/builtins/#url

    In Django 1.5, the behavior of the url template tag will change, with the
    first argument being made into a context variable, rather than being a
    special case unquoted constant. This will allow the url tag to use a
    context variable as the value of the URL name to be reversed.

    In order to provide a forwards compatibility path, Django 1.3 provides
    a future compatibility library -- future -- that implements the new behavior.
    To use this library, add a load call at the top of any template using the url
    tag, and wrap the first argument to the url tag in quotes. For example:

    {% load url from future %}

    {% url 'app_views.client' %}
    {% url 'myapp:view-name' %}

    {% with view_path="app_views.client" %}
    {% url view_path client.id %}
    {% endwith %}

    {% with url_name="client-detail-view" %}
    {% url url_name client.id %}
    {% endwith %}
"""
from __future__ import unicode_literals

import re
import os
import logging

log = logging.getLogger(__name__)

CURRENT_PATH = os.path.abspath('.')
load_tag = "{% load url from future %}"
file_formats = ['.html', '.txt']

re_flags = re.I | re.X | re.U

# Deprecated url tag
r_depr_url_finder = re.compile(r" {% \s* url \s+ [^\"\'\s]+ \s+ ", re_flags)

# And url tag
r_url_finder = re.compile(r"{% \s* url \s+ ", re_flags)
# {% load url from future %}
r_load_finder = re.compile(r" {% \s* load \s+ url \s+ from \s+ future \s* %} ", re_flags)
# {% extends ... %} tag
r_extends_finder = re.compile(r"{% \s* extends \s* \S+ \s* %}", re_flags)

# Modernizing url tag replace pattern
r_url_pattern = re.compile(r"""
    (?P<before> {% \s*
        url \s+ )
        (?P<name> \S+ )
        (?P<attrs> \s+ .*? )
    (?P<after> \s* %} )
""", re_flags)

# Add load from future after extends
r_load_extends_pattern = re.compile(
    r"(?P<template_head> [\s\S]* {% \s* extends \s* \S+ \s* %} )",
    re_flags | re.M,
)

# Adding load tag pattern
r_load_extends_replace = """\g<template_head>\n\n%s\n""" % load_tag


def make_me_magic(write):
    """ Main script.

    Here we find templates, replace old-style url tags and add future import where necessary.
    """
    # Search for files with appropriate extensions.
    log.info('Files needing modification:')
    found = False
    for file_path in find_files(os.walk(CURRENT_PATH)):
        with open(file_path, 'r+') as t_file:
            file_content = t_file.read()
            # Checking for presence of old-style tags and absence of load url from future
            if has_deprecated_tag(file_content):
                found = True
                log.info(file_path.replace(CURRENT_PATH + '/', ''))
                new_content = parse_file(file_content)

                if write:
                    t_file.seek(0)
                    t_file.write(new_content)
                    log.info('    File updated')

    if not found:
        log.info('All files are up to date. Congrats!')
    else:
        if not write:
            log.info('No actual changes made. Run future_url --write to fix files right now.')



def find_files(paths):
    for dirpath, _, filenames in paths:
        for path in search_template_files(dirpath, filenames):
            yield path


def parse_file(file_content):
    if not has_deprecated_tag(file_content):
        return file_content

    # Handle files with deprecated url tags
    if r_depr_url_finder.search(file_content):
        if check_comma_in_attributes(file_content):
            log.warn("    Comma separated attributes in url tag are no longer supported.")
        file_content = process_url_tag(file_content)

    # Check if load url form future is present and add if necessary
    if r_url_finder.search(file_content) and not r_load_finder.search(file_content):
        file_content = process_load_tag(file_content)
        log.info("    Need to add {% load url from future %}")

    return file_content


def url_replacer(match):
    """ Handle single url tag. """

    matches = match.groupdict()
    if ',' in match.group('attrs'):
        matches['attrs'] = re.sub('\s*,\s*', ' ', match.group('attrs'))
    repl = "{before}'{name}'{attrs}{after}".format(**matches)
    logging.info("    Proposed replace: {0} -> {1}".format(match.group(0), repl))
    return repl


def process_load_tag(html):
    """ Add {% load url from future %} """

    if r_extends_finder.search(html):
        return r_load_extends_pattern.sub(r_load_extends_replace, html, count=1)
    else:
        return "{load_tag}\n{html}".format(
            load_tag=load_tag,
            html=html,
        )


def process_url_tag(html):
    """ Modernize url tag. """

    return r_url_pattern.sub(url_replacer, html)


def check_comma_in_attributes(html):
    """ Attributes of url tag can't be separated by comma. """

    search = r_url_pattern.search(html)
    return ',' in search.group('attrs')


def has_deprecated_tag(html):
    """ Check if html needs modernizing. """

    # File is already up-to-date
    has_load_tag = r_load_finder.search(html)

    # No load tag and (deprecated) url tag present
    return not has_load_tag and (r_depr_url_finder.search(html) or r_url_finder.search(html))


def search_template_files(dirname, fnames):
    """ Search for suitable files """

    for file_name in fnames:
        if os.path.splitext(file_name)[1] in file_formats:
            yield os.path.join(dirname, file_name)

django_future_url/main.py
"""
Django future url.

Migration tool for django 1.4, fixes url template tag deprecation warnings.

    -  Finds all html``, txt files.
    -  Replaces all old style url tags.
    -  Inserts {% load url from future %} when it's needed.

Usage:
  future_url [options]
  future_url -h | --help
  future_url --version

Options:
  -w --write    Actually make replacements.
  -q --quiet    Do not print anything to stdout.
  -h --help     Show this screen.
  --version     Show version.
"""
import logging

from docopt import docopt
import django_future_url
from django_future_url.core import make_me_magic


def future_url():
    arguments = docopt(__doc__, version='Django future url ' + django_future_url.__version__)

    # Set up logging handler
    level = logging.CRITICAL if arguments['--quiet'] else logging.INFO
    logging.basicConfig(format='%(message)s', level=level)

    make_me_magic(write=arguments['--write'])


django_future_url/django_make_future_url.py
#!/usr/bin/env python
# coding: utf-8

""" Searches for templates in current directory recursively and
modernizes them to use new {% url %} syntax.
Adds {% load url from future %} on top of your template if necessary (Django 1.3-1.4).

From the Django docs
https://docs.djangoproject.com/en/1.4/ref/templates/builtins/#url

    In Django 1.5, the behavior of the url template tag will change, with the
    first argument being made into a context variable, rather than being a
    special case unquoted constant. This will allow the url tag to use a
    context variable as the value of the URL name to be reversed.

    In order to provide a forwards compatibility path, Django 1.3 provides
    a future compatibility library -- future -- that implements the new behavior.
    To use this library, add a load call at the top of any template using the url
    tag, and wrap the first argument to the url tag in quotes. For example:

    {% load url from future %}

    {% url 'app_views.client' %}
    {% url 'myapp:view-name' %}

    {% with view_path="app_views.client" %}
    {% url view_path client.id %}
    {% endwith %}

    {% with url_name="client-detail-view" %}
    {% url url_name client.id %}
    {% endwith %}»
"""

import re
import os
from optparse import OptionParser
import django_future_url

CURRENT_PATH = os.path.abspath('.')
template_files = []
load_tag = "{% load url from future %}"
file_formats = ['.html', '.txt']

re_flags = re.I | re.X | re.U

# Deprecated url tag
r_depr_url_finder = re.compile(ur" {% \s* url \s+ [^\"\'\s]+ \s+ ", re_flags)

# And url tag
r_url_finder = re.compile(ur"{% \s* url \s+ ", re_flags)
# {% load url from future %}
r_load_finder = re.compile(ur" {% \s* load \s+ url \s+ from \s+ future \s* %} ", re_flags)
# {% extends ... %} tag
r_extends_finder = re.compile(ur"{% \s* extends \s* \S+ \s* %}", re_flags)

# Modernizing url tag replace pattern
r_url_pattern = re.compile(ur"""
    (?P<before> {% \s*
        url \s+ )
        (?P<name> \S+ )
        (?P<attrs> \s+ .*? )
    (?P<after> \s* %} )
""", re_flags)

# Add load from future after extends
r_load_extends_pattern = re.compile(
    ur"(?P<template_head> [\s\S]* {% \s* extends \s* \S+ \s* %} )",
    re_flags | re.M,
)

# Adding load tag pattern
r_load_extends_replace = """\g<template_head>\n\n%s\n""" % load_tag

# Options
parser = OptionParser("usage: %prog [options]", version='%prog ' + django_future_url.__version__)
parser.add_option(
    "-v", "--verbose",
    action="store_true",
    dest="verbose",
    default=False,
    help="Print status messages to stdout"
)
parser.add_option(
    "-D", "--dry-run",
    dest="dryrun",
    action="store_true",
    default=False,
    help='Only shows changes to be made without actually modifying files'
)
(options, args) = parser.parse_args()


def make_me_magic():
    """ Main script.

    Here we find templates, replace old-style url tags and add future import where necessary.
    """

    # Search for files with appropriate extensions.
    os.path.walk(CURRENT_PATH, search_template_files, template_files)

    for file_path in template_files:
        with open(file_path, 'r+') as t_file:
            file_content = t_file.read()
            print(file_content)
            new_content = parse_file(file_content)
            if new_content != file_content and not getattr(options, 'dryrun', True):
                t_file.seek(0)
                t_file.write(new_content)
                message("File updated")
            message('\n')


def parse_file(file_content):
    # Checking for presence of old-style tags and absence of load url from future
    if has_deprecated_tag(file_content):
        # print t_file.name

        # Handle files with deprecated url tags
        if r_depr_url_finder.search(file_content):
            # Comma separated attributes are no longer supported in load from future
            if check_comma_in_attributes(file_content):
                message("Comma separated attribudes in “future” url tag are not supported.")
            file_content = process_url_tag(file_content)

        # Check if load url form future is present and add if necessary
        if r_url_finder.search(file_content) and not r_load_finder.search(file_content):
            file_content = process_load_tag(file_content)
            message("Added {% load url from future %}")

    return file_content


def url_replacer(match):
    """ Handle single url tag. """

    matches = match.groupdict()
    if ',' in match.group('attrs'):
        matches['attrs'] = re.sub('\s*,\s*', ' ', match.group('attrs'))
    repl = "{before}'{name}'{attrs}{after}".format(**matches)
    message(u"replaced: {0} -> {1}".format(match.group(0), repl))
    return repl


def process_load_tag(html):
    """ Add {% load url from future %} """

    if r_extends_finder.search(html):
        return r_load_extends_pattern.sub(r_load_extends_replace, html, count=1)
    else:
        return "{load_tag}\n{html}".format(
            load_tag=load_tag,
            html=html,
        )


def process_url_tag(html):
    """ Modernize url tag. """

    return r_url_pattern.sub(url_replacer, html)


def check_comma_in_attributes(html):
    """ Attributes of load tag can't be separated by comma. """

    search = r_url_pattern.search(html)
    return ',' in search.group('attrs')


def has_deprecated_tag(html):
    """ Check if html needs modernizing. """

    # File is already up-to-date
    has_load_tag = r_load_finder.search(html)

    # No load tag and (deprecated) url tag present
    return not has_load_tag and (r_depr_url_finder.search(html) or r_url_finder.search(html))


def search_template_files(template_files, dirname, fnames):
    """ Search for suitable files """

    for file_name in fnames:
        if os.path.splitext(file_name)[1] in file_formats:
            file_path = os.path.join(dirname, file_name)
            template_files.append(file_path)


def message(txt):
    if getattr(options, 'verbose', False) or getattr(options, 'dryrun', False):
        print txt


if __name__ == '__main__':
    make_me_magic()


django-future-url/django-make-future-url.py
#!/usr/bin/env python
# coding: utf-8

""" Searches for templates in current directory recursively and
modernizes them to use new {% url %} syntax.
Adds {% load url from future %} on top of your template if necessary (Django 1.3-1.4).

From the Django docs
https://docs.djangoproject.com/en/1.4/ref/templates/builtins/#url

    In Django 1.5, the behavior of the url template tag will change, with the
    first argument being made into a context variable, rather than being a
    special case unquoted constant. This will allow the url tag to use a
    context variable as the value of the URL name to be reversed.

    In order to provide a forwards compatibility path, Django 1.3 provides
    a future compatibility library -- future -- that implements the new behavior.
    To use this library, add a load call at the top of any template using the url
    tag, and wrap the first argument to the url tag in quotes. For example:

    {% load url from future %}

    {% url 'app_views.client' %}
    {% url 'myapp:view-name' %}

    {% with view_path="app_views.client" %}
    {% url view_path client.id %}
    {% endwith %}

    {% with url_name="client-detail-view" %}
    {% url url_name client.id %}
    {% endwith %}»
"""

import re
import os
from optparse import OptionParser

version = __import__('django-future-url').get_version()

CURRENT_PATH = os.path.abspath('.')
template_files = []
load_tag = "{% load url from future %}"
file_formats = ['.html', '.txt']

re_flags = re.I | re.X | re.U

# Deprecated url tag
r_depr_url_finder = re.compile(ur" {% \s* url \s+ [^\"\'\s]+ \s+ ", re_flags)

# And url tag
r_url_finder = re.compile(ur"{% \s* url \s+ ", re_flags)
# {% load url from future %}
r_load_finder = re.compile(ur" {% \s* load \s+ url \s+ from \s+ future \s* %} ", re_flags)
# {% extends ... %} tag
r_extends_finder = re.compile(ur"{% \s* extends \s* \S+ \s* %}", re_flags)

# Modernizing url tag replace pattern
r_url_pattern = re.compile(ur"""
    (?P<before> {% \s*
        url \s+ )
        (?P<name> \S+ )
        (?P<attrs> \s+ .*? )
    (?P<after> \s* %} )
""", re_flags)

# Add load from future after extends
r_load_extends_pattern = re.compile(
    ur"(?P<template_head> [\s\S]* {% \s* extends \s* \S+ \s* %} )",
    re_flags | re.M,
)

# Adding load tag pattern
r_load_extends_replace = """\g<template_head>\n\n%s\n""" % load_tag

# Options
parser = OptionParser("usage: %prog [options]", version='%prog ' + version)
parser.add_option(
    "-v", "--verbose",
    action="store_true",
    dest="verbose",
    default=False,
    help="Print status messages to stdout"
)
parser.add_option(
    "-D", "--dry-run",
    dest="dryrun",
    action="store_true",
    default=False,
    help='Only shows changes to be made without actually modifying files'
)
(options, args) = parser.parse_args()


def make_me_magic():
    """ Main script.

    Here we find templates, replace old-style url tags and add future import where necessary.
    """

    # Search for files with appropriate extensions.
    os.path.walk(CURRENT_PATH, search_template_files, template_files)

    for file_path in template_files:
        with open(file_path, 'r+') as t_file:
            file_content = t_file.read()

            # Checking for presence of old-style tags and absence of load url from future
            if has_deprecated_tag(file_content):
                print file_path

                # Handle files with deprecated url tags
                if r_depr_url_finder.search(file_content):
                    # Comma separated attributes are no longer supported in load from future
                    if check_comma_in_attributes(file_content):
                        message("Comma separated attribudes in “future” url tag are not supported.")
                    file_content = process_url_tag(file_content)

                # Check if load url form future is present and add if necessary
                if r_url_finder.search(file_content) and not r_load_finder.search(file_content):
                    file_content = process_load_tag(file_content)
                    message("Added {% load url from future %}")

                if not getattr(options, 'dryrun', True):
                    t_file.seek(0)
                    t_file.write(file_content)
                    message("File updated")

                message('\n')


def url_replacer(match):
    """ Handle single url tag. """

    matches = match.groupdict()
    if ',' in match.group('attrs'):
        matches['attrs'] = re.sub('\s*,\s*', ' ', match.group('attrs'))
    repl = "{before}'{name}'{attrs}{after}".format(**matches)
    message(u"replaced: {0} -> {1}".format(match.group(0), repl))
    return repl


def process_load_tag(html):
    """ Add {% load url from future %} """

    if r_extends_finder.search(html):
        return r_load_extends_pattern.sub(r_load_extends_replace, html, count=1)
    else:
        return "{load_tag}\n{html}".format(
            load_tag=load_tag,
            html=html,
        )


def process_url_tag(html):
    """ Modernize url tag. """

    return r_url_pattern.sub(url_replacer, html)


def check_comma_in_attributes(html):
    """ Attributes of load tag can't be separated by comma. """

    search = r_url_pattern.search(html)
    return ',' in search.group('attrs')


def has_deprecated_tag(html):
    """ Check if html needs modernizing. """

    # File is already up-to-date
    has_load_tag = r_load_finder.search(html)

    # No load tag and (deprecated) url tag present
    return not has_load_tag and (r_depr_url_finder.search(html) or r_url_finder.search(html))


def search_template_files(template_files, dirname, fnames):
    """ Search for suitable files """

    for file_name in fnames:
        if os.path.splitext(file_name)[1] in file_formats:
            file_path = os.path.join(dirname, file_name)
            template_files.append(file_path)


def message(txt):
    if getattr(options, 'verbose', False) or getattr(options, 'dryrun', False):
        print txt


if __name__ == '__main__':
    make_me_magic()
